core/src/main/scala/org/scalafmt/internal/Router.scala

```diff
@@ -420,67 +420,121 @@ class Router(formatOps: FormatOps) {
+        val isBracket = open.is[LeftBracket]
 
         @tailrec
-        def loop(token: Token): Option[Token] =
+        def loop(token: Token): Option[Token] = {
           leftTok2tok(matchingParentheses(hash(token))) match {
-            case FormatToken(RightParen(), l @ LeftParen(), _) => loop(l)
-            case FormatToken(r @ RightParen(), _, _) => Some(r)
+            case FormatToken(RightParenOrBracket(), l @ LeftParen(), _) =>
+              loop(l)
+            // This case only applies to classes
+            case f @ FormatToken(RightBracket(), CtorModifier(), _) =>
+              loop(next(f).right)
+            case FormatToken(r @ RightParenOrBracket(), _, _) => Some(r)
             case _ => None
           }
-        // find the last param on the def
-        // so that we can apply our `policy` till
-        // the end.
+        }
+        // find the last param on the defn so that we can apply our `policy`
+        // till the end.
         val lastParen = loop(open).get
 
-        // If this is a class, we need to do some different things
-        // We only care about the Primary Ctor since other Ctors will
-        // be picked up by the `def` definition
-        val isCtor = leftOwner.is[meta.Ctor.Primary]
+        val mixedParams = {
+          leftOwner match {
+            case cls: meta.Defn.Class =>
+              cls.tparams.nonEmpty && cls.ctor.paramss.nonEmpty
+            case _ => false
+          }
+        }
+
+        val shouldNotDangle =
+          leftOwner.is[meta.Ctor.Primary] ||
+          leftOwner.is[meta.Defn.Class] ||
+          leftOwner.is[meta.Defn.Trait]
+
+        // Since classes and defs aren't the same (see below), we need to
+        // create two (2) OneArgOneLineSplit when dealing with classes. One
+        // deals with the type params and the other with the data params.
+        val oneLinePerArg = {
+          val base = OneArgOneLineSplit(open)
+          if (mixedParams) {
+            val afterTypes = leftTok2tok(matchingParentheses(hash(open)))
+            // Try to find the first paren. If found, then we are dealing with
+            // a class with type AND data params. Otherwise it is a class with
+            // just type params.
+            findFirst(afterTypes, lastParen)(t => t.left.is[LeftParen])
+              .fold(base)(t => base.merge(OneArgOneLineSplit(t.left)))
+          }
+          else base
+        }
+
+        // DESNOTE(2017-03-28, pjrt) Classes and defs aren't the same.
+        // For defs, type params and data param have the same `owners`. However
+        // this is not the case for classes. Type params have the class itself
+        // as the owner, but data params have the Ctor as the owner, so a
+        // simple check isn't enough. Instead we check against the owner of the
+        // `lastParen` as well, which will be the same as the data param's
+        // owner.
+        val dataParamsOwner = owners(lastParen)
```

Comment:
Nitpick, I typically use "value" instead of "data" to describe runtime values. At least in Scala, the convention seems to be to talk about types vs. values.